pdf-workitem/src/main/java/org/jbpm/process/workitem/pdf/GeneratePDFWorkitemHandler.java

```diff
@@ -0,0 +1,145 @@
+                keywords = "pdf,generate,template,document,thymeleaf",
+                action = @WidAction(title = "Generate PDF document")
+        ))
+public class GeneratePDFWorkitemHandler extends AbstractLogOrThrowWorkItemHandler {
+
+    private static final Logger logger = LoggerFactory.getLogger(GeneratePDFWorkitemHandler.class);
+    private static final String RESULTS_VALUE = "PDFDocument";
+    private static final String UTF_8 = "UTF-8";
+    private String resultXHTML;
+
+    public void executeWorkItem(WorkItem workItem,
+                                WorkItemManager workItemManager) {
+
+        try {
+
+            RequiredParameterValidator.validate(this.getClass(),
+                                                workItem);
+
+            Map<String, Object> results = new HashMap<>();
+            Document templateDocument = (Document) workItem.getParameter("TemplateDocument");
+            Object data = workItem.getParameter("TemplateData");
+            String dataKey = (String) workItem.getParameter("TemplateDataKey");
+            String generateDocumentName = (String) workItem.getParameter("GeneratedDocumentName");
+
+            TemplateEngine templateEngine = new TemplateEngine();
+            StringTemplateResolver resolver = new StringTemplateResolver();
+            resolver.setTemplateMode(HTML);
+            resolver.setCacheable(false);
+            templateEngine.setTemplateResolver(resolver);
+
+            Context context = new Context();
+            context.setVariable(dataKey,
+                                data);
+
+            String renderedHtmlContent = templateEngine.process(new String(templateDocument.getContent(),
+                                                                           StandardCharsets.UTF_8),
+                                                                context);
+            resultXHTML = convertToXhtml(renderedHtmlContent);
+
+            ITextRenderer renderer = new ITextRenderer();
+            renderer.setDocumentFromString(resultXHTML);
+            renderer.layout();
+
+            File generatedPDFile = File.createTempFile(generateDocumentName,
+                                                       ".pdf",
+                                                       null);
+            FileOutputStream fos = new FileOutputStream(generatedPDFile);
+            renderer.createPDF(fos);
+
+            results.put(RESULTS_VALUE,
+                        generatedPDFile);
+
+            workItemManager.completeWorkItem(workItem.getId(),
+                                             results);
+        } catch (Exception e) {
+            logger.error(e.getMessage());
+            handleException(e);
+        }
+    }
+
+    public void abortWorkItem(WorkItem workItem,
+                              WorkItemManager manager) {
+    }
+
+    private String convertToXhtml(String html) throws UnsupportedEncodingException {
```

Comment:
What’s the purpose of this method?

Reply:
itext needs XHTML and not just plain html. this method we convert the template to xhtml so that it can be used to generate the pdf

Reply:
so it has to be always the xhtml for IText? didn't know about it ... kind of weird ... maybe we should expect users to already provide templates in xhtml to reduce number of dependencies, wdyt?

Reply:
that's true that people will add 3rd party dependencies but the rule should still be applied to have as few dependencies as possible on kjar level - due to potential conflicts with libraries - see the mailing list thread about google calendar work item ...